File: Product-Of-Array-Except-Self.cpp
class Solution {
public:
    vector<int> productExceptSelfNormal(vector<int>& nums) {
        int count = 0, product = 1;
        vector<int> ans;
        //count the number of 0's
        for (int ele : nums) {
            if (ele == 0) {
                count++;
            }
            if (count > 1) return vector<int>( nums.size(), 0);
        }

        if (count == 1) {
            int product_without_zero = 1;
            for(int ele : nums) {
                if (ele != 0) product_without_zero *= ele;
            }
            for(int ele : nums) {
                if (ele == 0) {
                    ans.push_back(product_without_zero);
                } else {
                    ans.push_back(0);
                }
            }
        } else {
            for(int ele : nums) {
                product *= ele;
            }
            for(int ele : nums) {
                ans.push_back(product/ele);
            }
        }
        return ans;
    }

    vector<int> productExceptSelfFpBp(vector<int>& nums) {
        ios_base::sync_with_stdio(false);
        int size = nums.size();
        vector<int> fp(size, 1), bp(size, 1), ans(size, 0);
        //populate the fp
        for( int i = 1 ; i < size ; i++ ) {
            fp[i] = fp[i-1] * nums[i-1];
        }
        //populate the bp
        for ( int i = size - 2 ; i >= 0 ; i-- ) {
            bp[i] = bp[i+1] * nums[i+1];
        }

        //calculate the product of both and return
        for ( int i = 0 ; i < size ; i++ ) {
            ans[i] = fp[i] * bp[i];
        }

        return ans;
    }

    vector<int> productExceptSelf(vector<int>& nums) {
        ios_base::sync_with_stdio(false);
        int size = nums.size(), curr_prod = 1;
        vector<int> ans(size, 1);

        for (int i = 0 ; i < size ; i++ ) {
            ans[i] *= curr_prod;
            curr_prod *= nums[i];
        }

        curr_prod = 1;
        for ( int i = size - 1 ; i >= 0 ; i-- ) {
            ans[i] *= curr_prod;
            curr_prod *= nums[i];
        }

        return ans;
    }
};

/*
Link: https://leetcode.com/problems/product-of-array-except-self/description/?envType=study-plan-v2&envId=top-interview-150

first solution:
Its simple. Count the number of 0's. If they are more than 1, that means any product except self will contain the other 0 resulting in the whole array having 0's. 
if the 0 count == 1, that means only 1 element is 0, and only at that index where we find 0 in the nums array, we will have a number having value of the product of the entire array except the 0 (product_except_0) and other places will be filled with 0's. 
and the last case is when there are no 0s where we simply calculate the product of the entire array and we populate the answer array with product / element (at that index).

second solution:
intuitive in the way that we consider a forward product and a backweard product. (fp and bp)
the idea behind this is that we need to have to find a way to store the products of each element BEFORE and AFTER a certain index cuz we dont want to include the element AT the INDEX. which is why we consider the previous index (for fp) and backward index (for bp)
simply put, at each index in FP, it contains the product of each element BEFORE that index and vice versa for the bp (product of each element AFTER that index)
The answer will be the product of the 2 arrays at each index respectively cuz finally we want product except self which is being achieved by the bp and fp (product BEFORE x product AFTER - both of which DONT contain the current elememnt at the index being calculated.)

third solution:
difficult to explain here but if you trace the execution of the second soltion, you will realise that it is iterating over the same arry 2 times front and back with repeated calcs. Just trace the answer and you will understand why we are doing this. 
curr_prod gets updated after every iteration with the prod of itself and the curr_number (nums[i]) and it is used to populate the next element in the ans[i]. 
Refer the last pages of your personal diary (aniketh) to see a small calc of the above. [nums[i-2]*nums[i-1]*nums[i]*prod]
*/

File: Product-Of-Array-Except-Self.java
class Solution {
    public int[] productExceptSelfWithArrays(int[] nums) {
        int counter = 0;
        int firstZeroIdx = -1;
        for(int i = 0 ; i < nums.length ; ++i) {
            if(nums[i] == 0) { 
                if(counter == 0) firstZeroIdx = i;
                counter++; 
            }
        }
        List<Integer> ans = new ArrayList<Integer>();
        if(counter == 1) {
            //one 0 means ans will be all 0's except where the idx is 0
            int prod = 1;
            for (int num : nums) {
                if(num != 0) prod *= num;
            }
            for( int i = 0 ; i < nums.length ; ++i ) {
                if(i == firstZeroIdx) ans.add(prod);
                else ans.add(0);
            }
        } else if (counter > 1) {
            // more than 1 0 means everywhere it will be 0
            for( int i = 0 ; i < nums.length ; ++i ) {
                ans.add(0);
            }
        } else {
            Integer[] forwardList = new Integer[nums.length];
            Integer[] backwardList = new Integer[nums.length];
            forwardList[0] = 1;
            backwardList[nums.length-1] = 1;
            int rollingProduct = 1;

            // fill in forward product
            for(int i = 1 ; i < nums.length ; ++i) {
                rollingProduct *= nums[i-1];
                forwardList[i] = rollingProduct;
            }

            rollingProduct = 1;
            //fill in backward product
            for(int i = nums.length - 2 ; i >= 0 ; --i) {
                rollingProduct *= nums[i+1];
                backwardList[i] = rollingProduct;
            }

            for(int i = 0 ; i < nums.length ; ++i) {
                ans.add(forwardList[i]*backwardList[i]);
            }

        }

        return ans.stream().mapToInt(Integer::intValue).toArray();
    }

    public int[] productExceptSelf(int[] nums) {
        //combining both left prod and right prod array into a single arr result and it passes in 2 goes to get the result.
        int[] result = new int[nums.length];
        result[0] = 1;
        for(int i = 1 ; i < nums.length ; ++i ) {
            result[i] = result[i-1] * nums[i-1];
        }

        int right = 1;
        for( int i = nums.length - 1 ; i >= 0 ; --i ) {
            result[i] *= right;
            right *= nums[i];
        }
        
        return result;
    }
}


/*
Algorithm Notes — productExceptSelfWithArrays & productExceptSelf

1. The goal: For each index, return the product of all other elements in the array
   without using division, handling cases with zeros carefully.

2. Step 1: Count zeros in the input array.
   - If more than one zero → every product will be 0.
   - If exactly one zero → only the position of that zero gets the product of all nonzero elements.
   - If no zeros → compute products normally.

3. Step 2 (no zeros case):
   - Build two helper arrays:
       forwardList[i] = product of all elements to the left of i.
       backwardList[i] = product of all elements to the right of i.
   - Multiply forwardList[i] * backwardList[i] to get the final product for each index.

4. Step 3 (optimized version — productExceptSelf):
   - Use a single result array.
   - First pass: fill result[i] with product of all elements to its left.
   - Second pass (right-to-left): multiply each result[i] by product of elements to its right.
   - This avoids extra arrays and runs in O(n) time with O(1) extra space.

5. Both approaches return an int[] where each element represents
   the product of all other numbers except the one at that index.
*/
